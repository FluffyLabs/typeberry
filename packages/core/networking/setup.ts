import QuicLogger, { formatting, LogLevel, StreamHandler } from "@matrixai/logger";
import { events, QUICClient, type QUICConnection, QUICServer, QUICSocket } from "@matrixai/quic";
import { BytesBlob } from "@typeberry/bytes";
import type { Ed25519Pair } from "@typeberry/crypto/ed25519.js";
import { Level, Logger } from "@typeberry/logger";
import {
  altNameRaw,
  certToPEM,
  ed25519AsJsonWebKeyPair,
  generateCertificate,
  type PeerInfo,
  privateKeyToPEM,
} from "./certificate.js";
import { getQuicClientCrypto, getQuicServerCrypto } from "./crypto.js";
import type { DialOptions } from "./network.js";
import { peerVerification } from "./peer-verification.js";
import { type PeerAddress, PeersManagement } from "./peers.js";
import { QuicNetwork } from "./quic-network.js";
import { QuicPeer } from "./quic-peer.js";
import { addEventListener } from "./quic-utils.js";

const logger = Logger.new(import.meta.filename, "net");

/** Networking server part options. */
export type Options = {
  /** Peer's ed25519key. */
  key: Ed25519Pair;
  /** Host interface to bound to. */
  host: string;
  /** Port to listen on. Use `0` for random. */
  port: number;
  /** Supported ALPNs (protocols) both for the inbound and outbound connections. */
  protocols: string[];
};

enum CloseReason {
  PeerIdMismatch = 0,
  DuplicateConnection = 1,
  ConnectionFromOurself = 2,
}

export class Quic {
  /** Setup QUIC socket and start listening for connections. */
  static async setup({ host, port, protocols, key }: Options): Promise<QuicNetwork> {
    const quicLoggerLvl = logger.getLevel() > Level.TRACE ? LogLevel.WARN : LogLevel.DEBUG;
    const quicLogger = new QuicLogger("quic", quicLoggerLvl, [
      new StreamHandler(formatting.format`${formatting.level}:${formatting.keys}:${formatting.msg}`),
    ]);

    // Load keypair
    const keyPair = ed25519AsJsonWebKeyPair(key);
    const privKeyPEM = await privateKeyToPEM(keyPair);
    const cert = await generateCertificate({
      certId: BytesBlob.blobFromString("QUIC Networking"),
      subjectKeyPair: keyPair,
      issuerKeyPair: keyPair,
    });

    const lastConnectedPeer = peerVerification();

    // QUICConfig
    const config = {
      keepAliveIntervalTime: 3000,
      maxIdleTimeout: 6000,
      applicationProtos: protocols,
      cert: certToPEM(cert),
      key: privKeyPEM,
      verifyPeer: true,
      verifyCallback: lastConnectedPeer.verifyCallback,
    };

    logger.info`🆔 Peer id: ** ${altNameRaw(key.pubKey)}@${host}:${port} ** (pubkey: ${key.pubKey})`;
    // Shared injected UDP socket
    const socket = new QUICSocket({
      logger: quicLogger.getChild("socket"),
    });

    // Start server on the socket.
    const server = new QUICServer({
      socket,
      config,
      crypto: getQuicServerCrypto(key),
      logger: quicLogger.getChild("server"),
    });

    // peer management
    const peers = new PeersManagement<QuicPeer>();

    // basic error handling
    addEventListener(server, events.EventQUICServerError, (error) => logger.error`🛜  Server error: ${error}`);
    addEventListener(server, events.EventQUICServerClose, (ev) => logger.error`🛜  Server stopped: ${ev}`);

    // handling incoming session
    addEventListener(server, events.EventQUICServerConnection, async (ev) => {
      const conn = ev.detail;
      if (lastConnectedPeer.info === null) {
        await conn.stop();
        return;
      }

      if (lastConnectedPeer.info.key.isEqualTo(key.pubKey)) {
        logger.log`🛜 Rejecting connection from ourself from ${conn.remoteHost}:${conn.remotePort}`;
        await conn.stop({ isApp: true, errorCode: CloseReason.ConnectionFromOurself });
        return;
      }

      if (peers.isConnected(lastConnectedPeer.info.id)) {
        logger.log`🛜 Rejecting duplicate connection with peer ${lastConnectedPeer.info.id} from ${conn.remoteHost}:${conn.remotePort}`;
        await conn.stop({ isApp: true, errorCode: CloseReason.DuplicateConnection });
        return;
      }

      logger.log`🛜 Server handshake with ${conn.remoteHost}:${conn.remotePort}`;
      newPeer(conn, lastConnectedPeer.info);
      lastConnectedPeer.info = null;
      await conn.start();
    });

    // connecting to a peer
    async function dial(peer: PeerAddress, options: DialOptions): Promise<QuicPeer> {
      return doDial();

      async function doDial() {
        const peerDetails = peerVerification();
        const clientLater = QUICClient.createQUICClient(
          {
            socket: socket,
            host: peer.host,
            port: peer.port,
            crypto: getQuicClientCrypto(),
            config: {
              ...config,
              verifyCallback: peerDetails.verifyCallback,
            },
            logger: quicLogger.getChild("client"),
          },
          {
            signal: options.signal,
          },
        );
        const client = await clientLater;

        if (peerDetails.info === null) {
          await client.destroy({ isApp: true, errorCode: CloseReason.PeerIdMismatch });
          throw new Error("Client connected, but there is no peer details!");
        }

        if (options.verifyName !== undefined && options.verifyName !== peerDetails.info.id) {
          await client.destroy({ isApp: true, errorCode: CloseReason.PeerIdMismatch });
          throw new Error(
            `Client connected, but the id didn't match. Expected: ${options.verifyName}, got: ${peerDetails.info.id}`,
          );
        }

        addEventListener(client, events.EventQUICClientClose, () => {
          logger.log`⚰️ Client connection closed.`;
        });

        addEventListener(client, events.EventQUICClientError, (error) => {
          logger.error`🔴 Client error: ${error.detail}`;
        });

        logger.log`🤝 Client handshake with: ${peer.host}:${peer.port}`;
        return newPeer(client.connection, peerDetails.info);
      }
    }

    function newPeer(conn: QUICConnection, peerInfo: PeerInfo) {
      const peer = new QuicPeer(conn, peerInfo);
      addEventListener(peer.conn, events.EventQUICConnectionClose, () => peers.peerDisconnected(peer));
      peers.peerConnected(peer);
      return peer;
    }

    return new QuicNetwork(socket, server, dial, peers, { host, port });
  }
}
