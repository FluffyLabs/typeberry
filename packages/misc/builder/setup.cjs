const fs = require("node:fs");
const childProcess = require("node:child_process");

const packageToBuild = process.env.PACKAGE_NAME;
let outDir = process.env.PACKAGE_OUT;

const isRelease = process.env.IS_RELEASE !== undefined;

if (!packageToBuild) {
  throw new Error(`Missing 'PACKAGE_NAME' environment variable.`);
}

if (!outDir) {
  const parts = packageToBuild.split("/");
  outDir = parts[parts.length - 1];
}

const DIST = `${__dirname}/../../../dist/${outDir}`;
const inputFile = `${__dirname}/pkg.ts`;

const data = `export * from "${packageToBuild}";`;
fs.writeFileSync(inputFile, data);

const commitHashResult = childProcess.execSync("git rev-parse --short HEAD");
const originalPackageJson = require(`${packageToBuild}/package.json`);
// generate package.json
const packageJson = JSON.stringify(
  {
    name: packageToBuild,
    version: isRelease
      ? originalPackageJson.version
      : `${originalPackageJson.version}-${commitHashResult.toString("utf8").trim()}`,
    main: "index.js",
    author: originalPackageJson.author,
    license: originalPackageJson.license,
    sideEffects: false,
    exports: {
      ".": {
        import: "./index.js",
        require: "./index.cjs",
      },
    },
  },
  null,
  2,
);
fs.mkdirSync(DIST, { recursive: true });
fs.writeFileSync(`${DIST}/package.json`, packageJson);

fs.writeFileSync(
  `${DIST}/.npmignore`,
  `
bin/**
packages/**
`,
);

module.exports = {
  inputFile,
  esmOutFile: `${DIST}/index.js`,
  cjsOutFile: `${DIST}/index.cjs`,
  typesInput: `${DIST}/packages/misc/builder/pkg.d.ts`,
};
