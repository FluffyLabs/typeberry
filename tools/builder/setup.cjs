const fs = require("node:fs");
const childProcess = require("node:child_process");

const packageToBuild = process.env.PACKAGE_NAME;
let outDir = process.env.PACKAGE_OUT;

if (!packageToBuild) {
  throw new Error(`Missing 'PACKAGE_NAME' environment variable.`);
}

if (!outDir) {
  const parts = packageToBuild.split("/");
  outDir = parts[parts.length - 1];
}

const data = `export * from "${packageToBuild}";`;
fs.writeFileSync(`${__dirname}/pkg.ts`, data);

const DIST = `${__dirname}/../../dist/${outDir}`;

const commitHashResult = childProcess.execSync("git rev-parse --short HEAD");
const originalPackageJson = require(`${packageToBuild}/package.json`);
// generate package.json
const packageJson = JSON.stringify(
  {
    name: packageToBuild,
    version: `${originalPackageJson.version}-${commitHashResult.toString("utf8").trim()}`,
    main: "index.js",
    author: originalPackageJson.author,
    license: originalPackageJson.license,
  },
  null,
  2,
);
fs.mkdirSync(DIST, { recursive: true });
fs.writeFileSync(`${DIST}/package.json`, packageJson);

fs.writeFileSync(
  `${DIST}/.npmignore`,
  `tools/**
packages/**
`,
);

module.exports = {
  outFile: `${DIST}/index.js`,
  typesInput: `${DIST}/tools/builder/pkg.d.ts`,
};
