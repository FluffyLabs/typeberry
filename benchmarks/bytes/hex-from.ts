import { add, complete, configure, cycle, save, suite } from "@typeberry/benchmark/setup";

function parseByteFromCharCodes(s: string): number {
  const a = parseCharCode(s.charCodeAt(0));
  const b = parseCharCode(s.charCodeAt(1));
  return (a << 8) | b;
}

const CODE_OF_0 = "0".charCodeAt(0);
const CODE_OF_9 = "9".charCodeAt(0);
const CODE_OF_a = "a".charCodeAt(0);
const CODE_OF_f = "f".charCodeAt(0);
const CODE_OF_A = "A".charCodeAt(0);
const CODE_OF_F = "F".charCodeAt(0);

function parseCharCode(x: number) {
  if (x >= CODE_OF_0 && x <= CODE_OF_9) {
    return x - CODE_OF_0;
  }

  if (x >= CODE_OF_a && x <= CODE_OF_f) {
    return 10 + x - CODE_OF_a;
  }

  if (x >= CODE_OF_A && x <= CODE_OF_F) {
    return 10 + x - CODE_OF_A;
  }

  throw new Error(`Invalid characters in hex byte string: ${String.fromCharCode(x)}`);
}

function parseByteFromNibbles(s: string): number {
  return (parseNibble(s[0]) << 4) | parseNibble(s[1]);
}
function parseNibble(n: string): number {
  switch (n) {
    case "0":
      return 0;
    case "1":
      return 1;
    case "2":
      return 2;
    case "3":
      return 3;
    case "4":
      return 4;
    case "5":
      return 5;
    case "6":
      return 6;
    case "7":
      return 7;
    case "8":
      return 8;
    case "9":
      return 9;
    case "A":
    case "a":
      return 10;
    case "B":
    case "b":
      return 11;
    case "C":
    case "c":
      return 12;
    case "D":
    case "d":
      return 13;
    case "E":
    case "e":
      return 14;
    case "F":
    case "f":
      return 15;
    default:
      throw new Error(`Invalid nibble: ${n}`);
  }
}

const data: string[] = [];
const size = 256;
for (let i = 0; i < size; i += 1) {
  data.push((i % 256).toString(16).padStart(2, "0"));
}

module.exports = () =>
  suite(
    "Bytes / hex parsing",

    add("parse hex using `Number` with NaN checking", () => {
      return data.map((byte) => {
        const n = Number(`0x${byte}`);
        if (Number.isNaN(n)) {
          throw new Error("Not a number");
        }
      });
    }),

    add("parse hex from char codes", () => {
      return data.map((byte) => parseByteFromCharCodes(byte));
    }),

    add("parse hex from string nibbles", () => {
      return data.map((byte) => parseByteFromNibbles(byte));
    }),

    cycle(),
    complete(),
    configure({}),
    ...save(__filename),
  );

if (require.main === module) {
  module.exports();
}
