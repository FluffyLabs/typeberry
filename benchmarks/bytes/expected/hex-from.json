{
  "name": "Bytes / hex parsing",
  "date": "2024-09-16T11:53:30.029Z",
  "version": null,
  "results": [
    {
      "name": "parse hex using `Number` with NaN checking",
      "ops": 134270.15,
      "margin": 0.1
    },
    {
      "name": "parse hex from char codes",
      "ops": 934650.55,
      "margin": 0.29
    },
    {
      "name": "parse hex from string nibbles",
      "ops": 556035.34,
      "margin": 0.25
    }
  ],
  "fastest": {
    "name": "parse hex from char codes",
    "index": 1
  },
  "slowest": {
    "name": "parse hex using `Number` with NaN checking",
    "index": 0
  }
}
